lucene/core/src/java/org/apache/lucene/search/MultiTermQueryConstantScoreWrapper.java

```diff
@@ -211,32 +218,39 @@ private WeightOrDocIdSet rewrite(LeafReaderContext context) throws IOException {
                 new ConstantScoreQuery(
                     new TermQuery(new Term(query.field, termsEnum.term()), termStates));
             Weight weight = searcher.rewrite(q).createWeight(searcher, scoreMode, score());
-            return new WeightOrDocIdSet(weight);
+            return new WeightOrDocIdSetIterator(weight);
           }
-          builder.add(docs);
+          PostingsEnum dropped = highFrequencyTerms.insertWithOverflow(postings);
+          otherTerms.add(dropped);
+          postings = dropped;
         } while (termsEnum.next() != null);
 
-        return new WeightOrDocIdSet(builder.build());
+        List<DocIdSetIterator> disis = new ArrayList<>(highFrequencyTerms.size() + 1);
+        for (PostingsEnum pe : highFrequencyTerms) {
+          disis.add(pe);
+        }
+        disis.add(otherTerms.build().iterator());
+        DisiPriorityQueue subs = new DisiPriorityQueue(disis.size());
+        for (DocIdSetIterator disi : disis) {
+          subs.add(new DisiWrapper(disi));
+        }
```

Comment:
Maybe I'm overlooking something silly, but can't we just do one pass like this?

```suggestion
DisiPriorityQueue subs = new DisiPriorityQueue(highFrequencyTerms.size() + 1);
        for (DocIdSetIterator disi : highFrequencyTerms) {
          subs.add(new DisiWrapper(disi));
        }
        subs.add(new DisiWrapper(otherTerms.build().iterator()));
```

Reply:
Also, it would be nice if we could get direct access to the underlying array backing `highFrequencyTerms`, then we could leverage `DisiPriorityQueue#addAll` to heapify everything at once.